Resource leak after calling Deno.readFileSync on directory

After calling Deno.readFileSync on a directory, a file handle stays open after the "Is a directory" error is thrown.

For example, take this test:

```ts
import { assertThrows } from "https://deno.land/std@0.92.0/testing/asserts.ts";

Deno.test("readFileSync on dir", async () => {
  assertThrows(() => (
    Deno.readFileSync('/home/paulo/misc/temp_dir')
  ), Error, "Is a directory");
});
```

After running it, an AssertionError is raised, because the handle is still open:

```
AssertionError: Test case is leaking resources.
Before: {
  "0": "stdin",
  "1": "stdout",
  "2": "stderr"
}
After: {
  "0": "stdin",
  "1": "stdout",
  "2": "stderr",
  "3": "fsFile"
}

Make sure to close all open resource handles returned from Deno APIs before
finishing test case.
    at assert (deno:runtime/js/06_util.js:33:13)
    at resourceSanitizer (deno:runtime/js/40_testing.js:84:7)
    at async Object.exitSanitizer [as fn] (deno:runtime/js/40_testing.js:100:9)
    at async TestRunner.[Symbol.asyncIterator] (deno:runtime/js/40_testing.js:272:13)
    at async Object.runTests (deno:runtime/js/40_testing.js:347:22)
    at async file:///home/paulo/misc/test_readfilesync/$deno$test.ts:3:1
```

Why does this happen?

--- a/cli/tests/unit/read_file_test.ts
+++ b/cli/tests/unit/read_file_test.ts
@@ -77,3 +77,21 @@ unitTest({ perms: { read: true } }, function readFileSyncLoop(): void {
     Deno.readFileSync("cli/tests/fixture.json");
   }
 });
+
+unitTest(
+  { perms: { read: true } },
+  async function readFileDoesNotLeakResources(): Promise<void> {
+    const resourcesBefore = Deno.resources();
+    await assertThrowsAsync(async () => await Deno.readFile("cli"));
+    assertEquals(resourcesBefore, Deno.resources());
+  },
+);
+
+unitTest(
+  { perms: { read: true } },
+  function readFileSyncDoesNotLeakResources(): void {
+    const resourcesBefore = Deno.resources();
+    assertThrows(() => Deno.readFileSync("cli"));
+    assertEquals(resourcesBefore, Deno.resources());
+  },
+);

--- a/cli/tests/unit/read_text_file_test.ts
+++ b/cli/tests/unit/read_text_file_test.ts
@@ -69,3 +69,21 @@ unitTest({ perms: { read: true } }, function readTextFileSyncLoop(): void {
     Deno.readTextFileSync("cli/tests/fixture.json");
   }
 });
+
+unitTest(
+  { perms: { read: true } },
+  async function readTextFileDoesNotLeakResources(): Promise<void> {
+    const resourcesBefore = Deno.resources();
+    await assertThrowsAsync(async () => await Deno.readTextFile("cli"));
+    assertEquals(resourcesBefore, Deno.resources());
+  },
+);
+
+unitTest(
+  { perms: { read: true } },
+  function readTextFileSyncDoesNotLeakResources(): void {
+    const resourcesBefore = Deno.resources();
+    assertThrows(() => Deno.readTextFileSync("cli"));
+    assertEquals(resourcesBefore, Deno.resources());
+  },
+);

--- a/runtime/js/40_read_file.js
+++ b/runtime/js/40_read_file.js
@@ -7,32 +7,44 @@
 
   function readFileSync(path) {
     const file = openSync(path);
-    const contents = readAllSync(file);
-    file.close();
-    return contents;
+    try {
+      const contents = readAllSync(file);
+      return contents;
+    } finally {
+      file.close();
+    }
   }
 
   async function readFile(path) {
     const file = await open(path);
-    const contents = await readAll(file);
-    file.close();
-    return contents;
+    try {
+      const contents = await readAll(file);
+      return contents;
+    } finally {
+      file.close();
+    }
   }
 
   function readTextFileSync(path) {
     const file = openSync(path);
-    const contents = readAllSync(file);
-    file.close();
-    const decoder = new TextDecoder();
-    return decoder.decode(contents);
+    try {
+      const contents = readAllSync(file);
+      const decoder = new TextDecoder();
+      return decoder.decode(contents);
+    } finally {
+      file.close();
+    }
   }
 
   async function readTextFile(path) {
     const file = await open(path);
-    const contents = await readAll(file);
-    file.close();
-    const decoder = new TextDecoder();
-    return decoder.decode(contents);
+    try {
+      const contents = await readAll(file);
+      const decoder = new TextDecoder();
+      return decoder.decode(contents);
+    } finally {
+      file.close();
+    }
   }
 
   window.__bootstrap.readFile = {

